### DISCUSSION TITLE
app.close()

### DISCUSSION THREAD

**Discussion by e3dio** [uNetworking/uWebSockets.js#575] [Ideas]
Call ws.close() on all sockets so don't need to keep Set of sockets and loop through to close app, also calls us_listen_socket_close?

**Comment by lveneris** [comment]
I second this - I need it for testing in Jest. I really don't want to create a worker process and have to SIGKILL it, because creating a subprocess for a parallel HTTP server is hacky.

**Reply by e3dio** [reply]
Can loop through a map to close sockets:

```javascript
let id = 0;
const wsMap = new Map();

app.ws('/ws', {
   open: ws => {
      wsMap.set(ws.id = id++, ws);
   },
   close: ws => {
      wsMap.delete(ws.id);
   }
}).listen(port, s => listenSocket = s);

const shutdown= sig =>{
   if (listenSocket) uws.us_listen_socket_close(listenSocket);
   wsMap.forEach(ws => ws.close());
}
process.on('SIGINT', shutdown);
process.on('SIGTERM', shutdown);
```
